cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -220,9 +416,11 @@ std::vector<size_type> segmented_count_set_bits(bitmask_type const* bitmask,
   if (num_indices == 0) {
     return std::vector<size_type>{};
   } else if (bitmask == nullptr) {
-    std::vector<size_type> ret(num_indices / 2);
-    for (size_t i = 0; i < num_indices / 2; i++) {
-      ret[i] = indices_begin[2 * i + 1] - indices_begin[2 * i];
+    std::vector<size_type> ret(num_ranges, 0);
+    if (count_bits == count_bits_policy::SET_BITS) {
+      for (size_type i = 0; i < num_ranges; i++) {
+        ret[i] = indices_begin[2 * i + 1] - indices_begin[2 * i];
+      }
```

Comment:
An empty bitmask doesn't have any bits, so this behavior doesn't quite make sense to me. Now, if this function was `segmented_valid_count` then this would make sense, but that's not what this function is.

Reply:
I agree this could be improved. This behavior matches the previous implementation and both set/unset cases are explicitly tested ([CountSetBitsTest, NullMask](https://github.com/rapidsai/cudf/blob/91141042ac5ce5024975eb2eab63f916047e6b6a/cpp/tests/bitmask/bitmask_tests.cpp#L68-L78), [CountUnsetBitsTest, NullMask](https://github.com/rapidsai/cudf/blob/91141042ac5ce5024975eb2eab63f916047e6b6a/cpp/tests/bitmask/bitmask_tests.cpp#L266-L276)), so I didn't want to change it. The [previous documentation](https://github.com/rapidsai/cudf/blob/91141042ac5ce5024975eb2eab63f916047e6b6a/cpp/include/cudf/detail/null_mask.cuh#L283) (``Returns an empty vector if `bitmask == nullptr`.``) might align with your expectation, but it disagreed with the implementation and tests. Do you have thoughts on how to resolve this? Would changing the name to `segmented_valid_count` be enough to address the ambiguity?

Reply:
Yuck, that's annoying. Is anything actually depending on this behavior? I'd rather see a new `segmented_valid/null_count` function built on top of `segmented_count_(un)set` bits that has this behavior of returning the size of each segment when the input mask is null.

Reply:
I like that idea. I would note that we already have a very large number of similarly named and similarly behaving APIs for the task of “segment bit counts” because of the various iterator types, host/device indices, set/unset counts, public/detail, and now “set vs. valid.” The intentions and scope of each existing (and proposed) function are well defined, but we will want good function docstrings and perhaps careful naming to help keep things straight.

Reply:
`count_set_bits` and `count_unset_bits` return 0 for a null bitmask: https://github.com/rapidsai/cudf/blob/31f92d70bdf5b80434d24b6883e45496dbe85042/cpp/include/cudf/null_mask.hpp#L103-L133

I think it would be most consistent to return a vector of 0s for the segmented versions.

Reply:
Also, we only need to add `segmented_valid_count` if the existing behavior is actually relied upon somewhere.

Reply:
The primitives "(segmented) count set/unset bits" are never used outside the context of valid/null bits. I am inclined to say that we should just change the names "(segmented) count set/unset bits" to "(segmented) count valid/null bits" and slightly modify the semantics to handle the case of `bitmask==nullptr` more appropriately.

Here's what I found while reviewing the codebase:

1. There are _zero_ uses of `count_set_bits` as a standalone function (only in the [context of `count_unset_bits`](https://github.com/rapidsai/cudf/blob/27b7190190adcc21f6a5e7a23249df8fddac1ba1/cpp/src/bitmask/null_mask.cu#L337)), so its default behavior of returning 0 when called with `bitmask==nullptr` is never triggered outside of tests.
2. The _few_ uses of `count_unset_bits` (almost all in [column_view.cpp](https://github.com/rapidsai/cudf/blob/27b7190190adcc21f6a5e7a23249df8fddac1ba1/cpp/src/column/column_view.cpp#L68-L70)) are all in the context of null counts (note, the semantics of unset counts are identical to null counts, even if `bitmask==nullptr`).
3. The _one_ use of `segmented_count_set_bits` ([orc/writer_impl.cu](https://github.com/rapidsai/cudf/blob/27b7190190adcc21f6a5e7a23249df8fddac1ba1/cpp/src/io/orc/writer_impl.cu#L903)) is in the context of null counts, and it must represent validity (so `bitmask==nullptr` needs a special case to return segment lengths).
4. The _one_ use of `segmented_count_unset_bits` ([slice.cu](https://github.com/rapidsai/cudf/blob/991136c78be01d4de20387086a185cfd5a21713b/cpp/src/copying/slice.cu#L43-L44)) is in the context of null counts (note, the semantics of unset counts are identical to null counts, even if `bitmask==nullptr`).

I investigated the idea @jrhemstad suggested of creating new functions like `segmented_valid_count` but the input validation / error checks are somewhat nontrivial and would need to be duplicated if `segmented_valid_count` were a wrapper for `segmented_count_set_bits`. Moreover, the behavior of `segmented_null_count` would be _exactly_ the same as `segmented_count_unset_bits` and that felt too redundant (identical semantics with unique names could be confusing to API consumers).

I propose the following:
- Rename `count_set_bits` to `count_valid_bits` and modify its behavior to return the total number of bits if `bitmask == nullptr`.
- Rename `count_unset_bits` to `count_null_bits` and continue to return 0 if `bitmask == nullptr` (unchanged semantics).
- Rename `segmented_count_set_bits` to `segmented_count_valid_bits` and continue to return a vector of segment lengths if `bitmask == nullptr` (unchanged semantics).
- Rename `segmented_count_unset_bits` to `segmented_count_null_bits` and continue to return a vector of zeros if `bitmask == nullptr` (unchanged semantics).

Summary: Since the code is _always_ dealing with validity/nullity bitmasks and never "set/unset bits" in any abstract form, I think we'd be best served by rename the functions for "set/unset bit counts" to eliminate their unclear semantics (e.g. `count_set_bits` not aligning with `segmented_count_set_bits` in the current implementation when `bitmask==nullptr`). Instead, we should name these APIs "(segmented) valid/null counts", the thing we actually care about in _all_ cases where these are used (see below), and slightly modify the semantics of one function (`count_set_bits`) to match our requirements in the case where `bitmask==nullptr`.

@jrhemstad @mythrocks @vyasr Curious to hear your thoughts - I implemented this in a3e7471b1078cd2377c7e5b2d1c58f33f081bbb4 so you can see what it would look like.

Reply:
> ... violates the single responsibility principle. count_set_bits doesn't know or care about the concept of valid/nulls... 

Appreciated. However, it does appear that all four of the functions @bdice listed here are used in the context of a validity buffers. While we might want support for `BOOL1` eventually, would it not be best to align the semantics of the APIs for the existing use cases? IMHO, renaming these functions to embrace the "null" semantics makes sense. 

We could consider adding the (semantically different) `count_set_bits()` back when its use case presents itself. (Apologies for the dissenting opinion.)

Reply:
If we really want to separate the concepts of validity and bitmasks cleanly, it seems like we would want to leave all of these methods as `*_bit_*` functions and then wrap them with corresponding `*_valid_*` functions that handle any logic specific to null masks. I view the fact that we represent validity with a bitmask (rather than e.g. a boolean bytemask) is part of the API, not as an implementation detail, so I don't necessarily think there's a problem applying a `*_bit_*` function directly to a validity mask. That being the case, we may not need to wrap all four of these methods, only the ones where we need special handling for e.g. a null bitmask.

Reply:
I discussed this offline with @jrhemstad earlier today and we concluded on a design that we both like. The plan is close to what @vyasr is describing, where "bits" functions are wrapped in "valid" functions.

- `count_set_bits` and its variants (`count_unset_bits`, `segmented_count_set_bits`, `segmented_count_unset_bits`) will all be detail methods and will raise if the bitmask is `nullptr`. These are purely for counting bits, and make no assumptions about the behavior of validity masks (hence raising instead of returning zero when receiving `nullptr`).
- New functions `valid_count`, `null_count`, `segmented_valid_count`, `segmented_null_count` will be the primary entry point for working with validity bitmasks. Existing calls of the "bits" functions will be updated to use these functions instead. All of these will be detail methods as well. Each of these functions will handle the assumptions of a fully-valid bitmask if `bitmask==nullptr`, and otherwise return the result of the corresponding "bits" function.
- The segmented bits functions will have documented requirements on their inputs (i.e. an even number of indices, indices must have `end >= start` for each range, etc.). These requirements will be enforced by the "valid" functions before passing inputs to the "bits" functions.

Consequences:
- All the APIs will be detail (removal of public API, so it is breaking).
- Assumptions about `bitmask==nullptr` will be handled via `valid`/`null` functions, rather than bit-counting functions.
- Future work can leverage bit counting APIs for other tasks, with fewer assumptions encoded into the logic.